pydra/engine/tests/test_submitter.py

```diff
@@ -575,3 +575,43 @@ def test_sge_no_limit_maxthreads(tmpdir):
         out_job2_dict["start_time"][0], f"%a %b %d %H:%M:%S %Y"
     )
     assert job_1_endtime > job_2_starttime
+
+
+def test_wf_with_blocked_tasks(tmpdir):
+    wf = Workflow(name="wf_with_blocked_tasks", input_spec=["x"])
+    wf.add(identity(name="taska", x=wf.lzin.x))
+    wf.add(alter_input(name="taskb", x=wf.taska.lzout.out))
+    wf.add(to_tuple(name="taskc", x=wf.taska.lzout.out, y=wf.taskb.lzout.out))
+    wf.set_output([("out", wf.taskb.lzout.out)])
+
+    wf.inputs.x = A(1)
+
+    wf.cache_dir = tmpdir
+
+    with pytest.raises(Exception, match="graph is not empty,"):
+        with Submitter("serial") as sub:
+            sub(wf)
+
+
+class A:
+    def __init__(self, a):
+        self.a = a
+
+    def __hash__(self):
+        return hash(self.a)
+
+
+@mark.task
+def identity(x):
+    return x
+
+
+@mark.task
+def alter_input(x):
+    x.a = 2
```

Comment:
Interesting reproduction. Here's a thought for how we could change this to make this use case smoother:

User writes a function that modifies its inputs. After execution, the runner re-checks the input hash and raises an error on change, rather than waiting for a workflow deadlock. The user then re-marks with `@mark.unsafe_task`, which causes the inputs to be `deepcopy`ed before execution. The input hash remains unchanged, and the workflow continues.

This doesn't resolve the hash nondeterminism. That's caused by https://github.com/nipype/pydra/blob/b5fe4c0eb7f937e70db15bc087d86fe90f401ff3/pydra/engine/helpers.py#L672-L674

Given that it's based on the string representation, you could probably resolve that on your end by sorting the set in `__str__` or `__repr__`. 

That said, there's a [StackOverflow](https://softwareengineering.stackexchange.com/questions/405243/how-to-perform-consistent-hashing-on-any-python-object-that-works-with-hash) on this exact problem, and the answer is basically that there's no way to hash arbitrary objects. But we could use `functools.singledispatch` to pass this problem off to users:

```Python
@functools.singledispatch
def hash_obj(obj: object) -> bytes:
    # Works for generic objects with __dict__
    dict_rep = ":".join(":".join(key, hash_obj(val)) for key, val in obj.__dict__.items())
    return sha256(f"{obj.__class__}:{dict_rep}".encode()).hexdigest() 

# Then define something for builtin classes with no __dict__
@hash_obj.register
def _(obj: int) -> bytes: ...

@hash_obj.register
def _(obj: str) -> bytes: ...

@hash_obj.register
def _(obj: dict) -> bytes: ...
```

We could even handle things like `set()` and `frozenset()` by default, and so only types that were really obscure would need to register their own. And we could either just let them decorate:


```Python
@pydra.util.hash_obj.register
def _(obj: File) -> bytes:
    ...
```

or provide a registration function

```Python
pydra.utils.register_hash(File, myhashfun)
```